### PR TITLE
Fixes for regexp replacement with backreferences (e.g. "\0")

### LINKED ISSUE #254: Regexp replace cannot use "\1"
#### Expected behaviour
regexp replace with "\1" should use the first subpartition

#### Actual behaviour
regexp replace with "\1" uses string "\1"

#### MATE general version
mate-desktop-1.18.0-5.fc26.x86_64

#### Package version
pluma-1.18.2-1.fc26.x86_64

#### Linux Distribution
Fedora 26

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/49949794-regexp-replace-cannot-use-1?utm_campaign=plugin&utm_content=tracker%2F1179776&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1179776&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by barkovv**
resolves #254 

Seems to work correct. Some additional tests from @hanaokaiwa are appreciated.

### CODE REVIEW

**Comment by barkovv** [comment]
@monsta I've removed all whitespace changes. This PR fixes backreferences either in search mode or in replace & replace all.

I've made some more tests and this code works as expected.

**Comment by barkovv** [comment]
@monsta are here any other problems with this PR? Can it be merged?

File: pluma/pluma-utils.c [hunk 1]
```diff
@@ -1636,66 +1637,58 @@ pluma_gtk_text_iter_regex_search (const GtkTextIter *iter,
...
-    found = g_regex_match_all (regex,text,0,&match_info);
-	if (found)
+	result_size = (gint) g_strv_length (all_matches);
+	non_null_result_number = (forward_search) ? 0 : (result_size -1);
+
+	non_null_result_found = FALSE;
+	while((non_null_result_number >= 0)
+	      && (non_null_result_number < result_size) )
 	{
-        all_matches = g_match_info_fetch_all (match_info);
```
Comment (monsta) [inline pluma/pluma-utils.c]: This part seems to be mostly about changing indents...

File: pluma/pluma-commands-search.c [hunk 2]
```diff
@@ -144,6 +144,11 @@ text_not_found (PlumaWindow *window,
 	g_free (searched);
 }
 
+/* This is the string with last replace statement. The reason to use some kind of
+   global variable is to keep existing methods' signatures not changed (to minimize
```
Comment (monsta) [inline pluma/pluma-commands-search.c]: This PR already has changes in 3 functions 🙂

**Comment by monsta** [comment]
I'll try various tests later

File: pluma/pluma-commands-search.c (hunk 2 above)
Comment (barkovv) [reply pluma/pluma-commands-search.c]: @monsta Three is less than six. Of course I can send this argument through all  sequence of calls if this global variable looks ugly. Or remove this comment.

File: pluma/pluma-commands-search.c [hunk 3]
```diff
@@ -271,7 +280,9 @@ do_find (PlumaSearchDialog *dialog,
 	}
 
 	g_free (search_text);
-	
+	if(match_regex)
+		last_replace_text = g_strdup(pluma_search_dialog_get_replace_text (dialog));
```
Comment (monsta) [inline pluma/pluma-commands-search.c]: I think `last_replace_text` should be freed before copying a new string to it.

File: pluma/pluma-commands-search.c [hunk 4]
```diff
@@ -404,20 +415,25 @@ do_replace (PlumaSearchDialog *dialog,
 
 	if (need_refind)
 	{
+		last_replace_text = g_strdup(replace_entry_text);
```
Comment (monsta) [inline pluma/pluma-commands-search.c]: Same here

File: pluma/pluma-commands-search.c (hunk 3 above)
Comment (barkovv) [reply pluma/pluma-commands-search.c]: @monsta fixed.

**Comment by barkovv** [comment]
@monsta I've moved global variable last_replace_text into PlumaDocument class (see last commit, just refactoring). I think it's ready to merge.

**Comment by monsta** [comment]
Yes, looks nicer now, will merge.

**Comment by sc0w** [comment]
We have a regression with this PR:

- The list of recently used files doesn't work

Without the merged commit, it works

**Comment by barkovv** [comment]
@sc0w I am not sure what "The list of recently used files doesn't work" means. Could you expand this statement a little please?

**Comment by sc0w** [comment]
click here, in the red area, and see the behavior with/without the PR:

g4520

**Comment by raveit65** [comment]
Any News for this regression?

**Comment by sc0w** [comment]
If the author doesn't fix it, I suggest to revert the culprit commit

**Comment by barkovv** [comment]
@raveit65 I've tried to launch both versions (with and without PR) and haven't found any issues. 

I need some detail steps to reproduce. Are you sure that this PR is guilty of your problem?

To be honest I think this PR doesn't touch anything else than search & replace logic and I can't understand how my changes may affect recent files list. Anyway if the regression have occured it must be fixed.

**Comment by monsta** [comment]
@sc0w: I can't reproduce it with git master. That button is working as before, it shows me the recently opened files.

**Comment by monsta** [comment]
> To be honest I think this PR doesn't touch anything else than search & replace logic and I can't understand how my changes may affect recent files list.

That's true, this PR doesn't affect that part of code.

**Comment by sc0w** [comment]
Reproducible steps:

1. rename or delete the file `/home/user/.local/share/recently-used.xbel`

2. install git master with the merged commit

3. touch testfile

4. kill all the pluma process and: pluma testfile

with_commit

5. install git master without the merged commit

without_commit

**Comment by raveit65** [comment]
@sc0w 
I followed your steps with generated pluma-1-19.3 from master from today, but i can' reproduce the issue.
pluma-test
pluma-test-2

**Comment by monsta** [comment]
I just got the same result. Tested in Debian Stretch (with all MATE stuff from git master).

**Comment by sc0w** [comment]
more info:

I can't reproduce the issue in debian stretch (VM)

but I can here, in debian testing

@lukefromdc please, can you test? (reproducible steps here: https://github.com/mate-desktop/pluma/pull/259#issuecomment-352280992)

**Comment by monsta** [comment]
Wow, I can reproduce it in my Debian Testing VM.
Really weird. I moved the original `recently-used.xbel` file to another folder before the test, and moved it back after it. Guess what, the issue isn't reproducible after the original file is moved back. 😕

**Comment by sc0w** [comment]
@monsta 

> Guess what, the issue isn't reproducible after the original file is moved back. 😕

You must have the issue using the existing .xbel file with these steps:

1. See the files existing when you click here:

g4520

2. open other files and verify, you always see the same files, no more files will be added to the list

**Comment by monsta** [comment]
It's weird, but I confirm this as well. The file itself is modified, new entries are added to it, its modification date/time changes, but... new entries aren't shown in Pluma. 😕

**Comment by monsta** [comment]
I also confirm that reverting this PR makes new entries appear in Pluma.

The entries that were saved with this PR lack some data that is normally present.

"Good" entries:
```xml
  <bookmark href="file:///home/monsta/1/indicator-application/debian/control" added="2017-11-29T13:43:45Z" modified="2017-11-29T13:43:54Z" visited="2017-11-29T13:43:45Z">
    <info>
      <metadata owner="http://freedesktop.org">
        <mime:mime-type type="text/plain"/>
        <bookmark:groups>
          <bookmark:group>pluma</bookmark:group>
        </bookmark:groups>
        <bookmark:applications>
          <bookmark:application name="Pluma" exec="&apos;pluma %u&apos;" modified="2017-11-29T13:43:54Z" count="2"/>
        </bookmark:applications>
      </metadata>
    </info>
  </bookmark>
  <bookmark href="file:///home/monsta/1/indicator-application/data/indicator-application.desktop.in" added="2017-11-29T13:58:03Z" modified="2017-11-29T14:14:59Z" visited="2017-11-29T13:58:03Z">
    <info>
      <metadata owner="http://freedesktop.org">
        <mime:mime-type type="text/plain"/>
        <bookmark:groups>
          <bookmark:group>pluma</bookmark:group>
        </bookmark:groups>
        <bookmark:applications>
          <bookmark:application name="Pluma" exec="&apos;pluma %u&apos;" modified="2017-11-29T14:14:59Z" count="4"/>
        </bookmark:applications>
      </metadata>
    </info>
  </bookmark>
```

"Bad" entries:
```xml
  <bookmark href="file:///home/monsta/mate/pluma/configure.ac" added="2017-12-17T19:30:15Z" modified="2017-12-20T09:12:53Z" visited="2017-12-17T19:30:15Z">
    <info>
      <metadata owner="http://freedesktop.org">
        <mime:mime-type type="text/plain"/>
        <bookmark:applications>
          <bookmark:application name="Pluma" exec="&apos;pluma %u&apos;" modified="2017-12-20T09:12:53Z" count="2"/>
        </bookmark:applications>
      </metadata>
    </info>
  </bookmark>
  <bookmark href="file:///home/monsta/mate/pluma/Makefile.am" added="2017-12-20T09:13:30Z" modified="2017-12-20T09:13:30Z" visited="2017-12-20T09:13:30Z">
    <info>
      <metadata owner="http://freedesktop.org">
        <mime:mime-type type="text/x-matlab"/>
        <bookmark:applications>
          <bookmark:application name="Pluma" exec="&apos;pluma %u&apos;" modified="2017-12-20T09:13:30Z" count="1"/>
        </bookmark:applications>
      </metadata>
    </info>
  </bookmark>
```

**Comment by monsta** [comment]
Looks like something is going wrong in the memory when saving a new recent entry. I've changed the way the data is stored in the memory in https://github.com/mate-desktop/pluma/pull/269 (stole it from gedit). Works for me in Debian Testing.

**Comment by monsta** [comment]
Alright, the fix from #269 is now tested and merged.

**Pull request merged**